src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -996,7 +996,10 @@ public int hashCode() {
                 copyLogs();
                 logsToCopy.put(node.getId(), 0L);
             }
-            node.addAction(new TimingAction());
+
+            if (node.getPersistentAction(TimingAction.class) == null) {
+                node.addAction(new TimingAction());
```

Comment:
Possible race condition—might need to be inside the `completed` lock.

Reply:
I don't *think* it's possible to invoke twice, because in workflow-cps we add the action before GraphListeners get a chance to fire on it. 

The plus side it is that adding the action twice only adds a tiny bit of storage waste -- we look for the first action matching the class type.